In [1]:
import pandas as pd
from datetime import datetime
import json

In [2]:
# Function to convert Id and timestamp structures to Python objects
def convert_objects(row):
    for key, value in row.items():
        if isinstance(value, dict) and "$oid" in value:
            row[key] = value["$oid"]
        # Convert Date in Unix timestamp format (milliseconds since the epoch)
        elif isinstance(value, dict) and "$date" in value:
            timestamp = value["$date"] / 1000  # Convert milliseconds to seconds
            row[key] = datetime.fromtimestamp(timestamp)
    return row

# Users Table Data Exploration

In [3]:
# Read and convert the JSON file
users_df = pd.read_json('users.json', lines=True)
users_df = users_df.apply(convert_objects, axis=1)
users_df.head()

_id  active             createdDate  \
0  5ff1e194b6a9d73a3a9f1052    True 2021-01-03 09:24:04.800   
1  5ff1e194b6a9d73a3a9f1052    True 2021-01-03 09:24:04.800   
2  5ff1e194b6a9d73a3a9f1052    True 2021-01-03 09:24:04.800   
3  5ff1e1eacfcf6c399c274ae6    True 2021-01-03 09:25:30.554   
4  5ff1e194b6a9d73a3a9f1052    True 2021-01-03 09:24:04.800   

                lastLogin      role signUpSource state  
0 2021-01-03 09:25:37.858  consumer        Email    WI  
1 2021-01-03 09:25:37.858  consumer        Email    WI  
2 2021-01-03 09:25:37.858  consumer        Email    WI  
3 2021-01-03 09:25:30.597  consumer        Email    WI  
4 2021-01-03 09:25:37.858  consumer        Email    WI

In [4]:
#checking data information
users_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495 entries, 0 to 494
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   _id           495 non-null    object        
 1   active        495 non-null    bool          
 2   createdDate   495 non-null    datetime64[ns]
 3   lastLogin     433 non-null    datetime64[ns]
 4   role          495 non-null    object        
 5   signUpSource  447 non-null    object        
 6   state         439 non-null    object        
dtypes: bool(1), datetime64[ns](2), object(4)
memory usage: 23.8+ KB


In [6]:
#checking count of null values
users_df.isnull().sum()

_id              0
active           0
createdDate      0
lastLogin       62
role             0
signUpSource    48
state           56
dtype: int64

In [5]:
#checking percent null values
users_df.isnull().mean()

_id             0.000000
active          0.000000
createdDate     0.000000
lastLogin       0.125253
role            0.000000
signUpSource    0.096970
state           0.113131
dtype: float64

In [7]:
#checking unique values and counts for the fields
print(users_df["_id"].nunique())
print(users_df["role"].value_counts())
print(users_df["signUpSource"].value_counts())
print(users_df['state'].value_counts())

212
role
consumer       413
fetch-staff     82
Name: count, dtype: int64
signUpSource
Email     443
Google      4
Name: count, dtype: int64
state
WI    396
NH     20
AL     12
OH      5
IL      3
KY      1
CO      1
SC      1
Name: count, dtype: int64


In [8]:
#checking for duplicated records
duplicate_records=users_df[users_df.duplicated()]
duplicate_records

_id  active             createdDate  \
1    5ff1e194b6a9d73a3a9f1052    True 2021-01-03 09:24:04.800   
2    5ff1e194b6a9d73a3a9f1052    True 2021-01-03 09:24:04.800   
4    5ff1e194b6a9d73a3a9f1052    True 2021-01-03 09:24:04.800   
5    5ff1e194b6a9d73a3a9f1052    True 2021-01-03 09:24:04.800   
8    5ff1e194b6a9d73a3a9f1052    True 2021-01-03 09:24:04.800   
..                        ...     ...                     ...   
490  54943462e4b07e684157a532    True 2014-12-19 08:21:22.381   
491  54943462e4b07e684157a532    True 2014-12-19 08:21:22.381   
492  54943462e4b07e684157a532    True 2014-12-19 08:21:22.381   
493  54943462e4b07e684157a532    True 2014-12-19 08:21:22.381   
494  54943462e4b07e684157a532    True 2014-12-19 08:21:22.381   

                  lastLogin         role signUpSource state  
1   2021-01-03 09:25:37.858     consumer        Email    WI  
2   2021-01-03 09:25:37.858     consumer        Email    WI  
4   2021-01-03 09:25:37.858     consumer        Email    WI  
5   2021-01-03 09:25:37.858     consumer        Email    WI  
8   2021-01-03 09:25:37.858     consumer        Email    WI  
..                      ...          ...          ...   ...  
490 2021-03-05 10:52:23.204  fetch-staff          NaN   NaN  
491 2021-03-05 10:52:23.204  fetch-staff          NaN   NaN  
492 2021-03-05 10:52:23.204  fetch-staff          NaN   NaN  
493 2021-03-05 10:52:23.204  fetch-staff          NaN   NaN  
494 2021-03-05 10:52:23.204  fetch-staff          NaN   NaN  

[283 rows x 7 columns]

In [9]:
#checking if there were any duplicated user_ids (with different information in other rows)
users_df[users_df['_id'].duplicated()]
# no additional duplicated user_ids

_id  active             createdDate  \
1    5ff1e194b6a9d73a3a9f1052    True 2021-01-03 09:24:04.800   
2    5ff1e194b6a9d73a3a9f1052    True 2021-01-03 09:24:04.800   
4    5ff1e194b6a9d73a3a9f1052    True 2021-01-03 09:24:04.800   
5    5ff1e194b6a9d73a3a9f1052    True 2021-01-03 09:24:04.800   
8    5ff1e194b6a9d73a3a9f1052    True 2021-01-03 09:24:04.800   
..                        ...     ...                     ...   
490  54943462e4b07e684157a532    True 2014-12-19 08:21:22.381   
491  54943462e4b07e684157a532    True 2014-12-19 08:21:22.381   
492  54943462e4b07e684157a532    True 2014-12-19 08:21:22.381   
493  54943462e4b07e684157a532    True 2014-12-19 08:21:22.381   
494  54943462e4b07e684157a532    True 2014-12-19 08:21:22.381   

                  lastLogin         role signUpSource state  
1   2021-01-03 09:25:37.858     consumer        Email    WI  
2   2021-01-03 09:25:37.858     consumer        Email    WI  
4   2021-01-03 09:25:37.858     consumer        Email    WI  
5   2021-01-03 09:25:37.858     consumer        Email    WI  
8   2021-01-03 09:25:37.858     consumer        Email    WI  
..                      ...          ...          ...   ...  
490 2021-03-05 10:52:23.204  fetch-staff          NaN   NaN  
491 2021-03-05 10:52:23.204  fetch-staff          NaN   NaN  
492 2021-03-05 10:52:23.204  fetch-staff          NaN   NaN  
493 2021-03-05 10:52:23.204  fetch-staff          NaN   NaN  
494 2021-03-05 10:52:23.204  fetch-staff          NaN   NaN  

[283 rows x 7 columns]

In [10]:
#counting duplicated records by user_id
duplicate_records.groupby('_id').count()

active  createdDate  lastLogin  role  signUpSource  \
_id                                                                            
54943462e4b07e684157a532      19           19         19    19             0   
59c124bae4b0299e55b0f330      17           17         17    17             0   
5a43c08fe4b014fd6b6a0612       7            7          7     7             0   
5fa41775898c7a11a6bcef3e      17           17         17    17            17   
5fb0a078be5fc9775c1f3945       1            1          0     1             1   
...                          ...          ...        ...   ...           ...   
60186237c8b50e11d8454d5f       4            4          0     4             4   
60189c74c8b50e11d8454eff       6            6          6     6             6   
60189c94c8b50e11d8454f6b       3            3          3     3             3   
601c2c05969c0b11f7d0b097       1            1          1     1             1   
60229990b57b8a12187fe9e0       1            1          1     1             1   

                          state  
_id                              
54943462e4b07e684157a532      0  
59c124bae4b0299e55b0f330     17  
5a43c08fe4b014fd6b6a0612      0  
5fa41775898c7a11a6bcef3e      0  
5fb0a078be5fc9775c1f3945      1  
...                         ...  
60186237c8b50e11d8454d5f      0  
60189c74c8b50e11d8454eff      6  
60189c94c8b50e11d8454f6b      3  
601c2c05969c0b11f7d0b097      1  
60229990b57b8a12187fe9e0      1  

[70 rows x 6 columns]

In [11]:
#creating a dataframe for unique user_ids
unique_users_df=users_df[~users_df.duplicated()]
unique_users_df

_id  active             createdDate  \
0    5ff1e194b6a9d73a3a9f1052    True 2021-01-03 09:24:04.800   
3    5ff1e1eacfcf6c399c274ae6    True 2021-01-03 09:25:30.554   
6    5ff1e1e8cfcf6c399c274ad9    True 2021-01-03 09:25:28.354   
7    5ff1e1b7cfcf6c399c274a5a    True 2021-01-03 09:24:39.626   
9    5ff1e1f1cfcf6c399c274b0b    True 2021-01-03 09:25:37.564   
..                        ...     ...                     ...   
435  5fc961c3b8cfca11a077dd33    True 2020-12-03 16:08:03.936   
455  5fa41775898c7a11a6bcef3e    True 2020-11-05 09:17:09.396   
456  5fa32b4d898c7a11a6bcebce    True 2020-11-04 16:29:33.309   
462  5964eb07e4b03efd0c0f267b    True 2017-07-11 10:13:11.771   
475  54943462e4b07e684157a532    True 2014-12-19 08:21:22.381   

                  lastLogin         role signUpSource state  
0   2021-01-03 09:25:37.858     consumer        Email    WI  
3   2021-01-03 09:25:30.597     consumer        Email    WI  
6   2021-01-03 09:25:28.392     consumer        Email    WI  
7   2021-01-03 09:24:39.665     consumer        Email    WI  
9   2021-01-03 09:25:37.599     consumer        Email    WI  
..                      ...          ...          ...   ...  
435 2021-02-26 16:39:16.799  fetch-staff        Email    NH  
455 2021-03-04 10:02:02.026  fetch-staff        Email   NaN  
456 2021-03-04 01:21:58.047  fetch-staff       Google    AL  
462 2021-03-04 13:07:49.770  fetch-staff          NaN    IL  
475 2021-03-05 10:52:23.204  fetch-staff          NaN   NaN  

[212 rows x 7 columns]

In [12]:
#checking count of null values
unique_users_df.isnull().sum()

_id              0
active           0
createdDate      0
lastLogin       40
role             0
signUpSource     5
state            6
dtype: int64

In [13]:
#checking percent null values
unique_users_df.isnull().mean()

_id             0.000000
active          0.000000
createdDate     0.000000
lastLogin       0.188679
role            0.000000
signUpSource    0.023585
state           0.028302
dtype: float64

In [14]:
#checking unique values and counts for the fields in the unique dataframe
print(unique_users_df["_id"].nunique())
print(unique_users_df["role"].value_counts())
print(unique_users_df["signUpSource"].value_counts())
print(unique_users_df['state'].value_counts())

212
role
consumer       204
fetch-staff      8
Name: count, dtype: int64
signUpSource
Email     204
Google      3
Name: count, dtype: int64
state
WI    193
AL      5
IL      3
KY      1
CO      1
OH      1
SC      1
NH      1
Name: count, dtype: int64


### Data Quality Issues: Users Table

1. Duplicated records: 

    There are 283 duplicated records, which is 57.2% of the data. This can have a large downstream impact when this data is used for analysis or marketing efforts
    
    Cleaning the duplicated records gives 212 unique records
    
2. Null values

    There's a small percentage of null values in the original data set in the following fields:
    
        lastLogin: 12.5%
        signUpSource: 9.7%
        state: 11.3%  
        
    Aftrer cleaning data set of duplicated records, the new percentages of null values are as follows:
    
        lastLogin: 18.9%
        signUpSource: 2.4%
        state: 2.8%  
        
    signUpSource and state will have minimal impact as there are very few null records. lastLogin may impact analysis that utilize this field.
    
    Reccomendations:  
        lastLogin:    
        
            The null values could be due to the user not logging after creating the account.
            The null records can be dropped or substituted with the createDate.  
            

        signUpSource:  

            The data distribution leans heavily to 'Email', so the null values can be substituted with 'Email' without impacting the data  
            

        state:  
        

            The data distribution leans heavily to 'WI', so the null values can be substituted with 'WI' without heavily impacting the data
